File: LifeCycleDemo/app/src/main/java/com/example/deakyu/lifecycledemo/MainActivity.java
package com.example.deakyu.lifecycledemo;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {

    private EditText editText;
    private TextView textView;
    private Button button;

    // LifeCycles are in order

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        editText = findViewById(R.id.edit_text);
        textView = findViewById(R.id.text_view);
        button = findViewById(R.id.button);

        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String value = editText.getText().toString();
                if(value != null) {
                    textView.setText(value);
                } else {
                    textView.setText("Value is null!");
                }
            }
        });

        System.out.println("DEE onCreate");
    }

    @Override
    protected void onStart() {
        super.onStart();

        System.out.println("DEE onStart");
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);

        String previousText = savedInstanceState.getString("currentText");
        textView.setText(previousText);

        System.out.println("DEE onRestoreInstanceState");
    }

    @Override
    protected void onResume() {
        super.onResume();

        System.out.println("DEE onResume");
    }

    @Override
    protected void onPause() {
        super.onPause();

        System.out.println("DEE onPause");
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        String currentText = textView.getText().toString();
        outState.putString("currentText", currentText);

        System.out.println("DEE onSaveInstanceState");
    }

    @Override
    protected void onStop() {
        super.onStop();

        System.out.println("DEE onStop");
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        System.out.println("DEE onDestroy");
    }
}

File: LifeCycleObservation/app/src/main/java/com/example/lee52/lifecycleobservation/MainActivity.kt
package com.example.lee52.lifecycleobservation

import android.content.Intent
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.Button

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        Log.d("DEAKYULEE", "onCreate MainActivity")

        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        findViewById<Button>(R.id.main_activity_button).setOnClickListener {
            val intent = Intent(MainActivity@this, NextActivity::class.java)
            startActivity(intent)
        }
    }

    override fun onStart() {
        super.onStart()
        Log.d("DEAKYULEE", "onStart MainActivity")
    }

    override fun onResume() {
        super.onResume()
        Log.d("DEAKYULEE", "onResume MainActivity")
    }

    override fun onPause() {
        super.onPause()
        Log.d("DEAKYULEE", "onPause MainActivity")
    }

    override fun onStop() {
        super.onStop()
        Log.d("DEAKYULEE", "onStop MainActivity")
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.d("DEAKYULEE", "onDestroy MainActivity")
    }
}
